### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by pablomendezroyo**
## Context
DAppNodeSDK is a critical dappnode tool used daily in Github actions. It has not received maintenance from a long time ago and some things are out of date.

Furthermore, there is code duplicated between dappmanager and the SDK:
- Docker-compose safe keys
- Manifest schema
- Setup wizard schema

This code duplication makes the developing experience annoying and results in production errors due to missed updates to both repositories

## Approach
The SDK should contain all the types and validating functions, then the dappmanager should use the DAppNodeSDK as a dependency and import these types and validating functions to be used in during the packages installation. This will benefit DAppNode in a better development experience, fewer bugs in production, easier implementation for new wizard/manifest/compose keys features, etc.

### Roadmap
- [x] Set types in the DAppNodeSDK (Take into account that the dappmanager joins all the release files, this will result in a manifest type with the added keys `setupWizard`, `gettingStarted`...)
- [x] Update any dependencies if required
   - ajv & ajv-errors
   - typescript 
- [x] Update Manifest schema and remove deprecated keywords
- [x] Implement schemas
- [x] Implement validating functions:
- [x] Implement tests for all the schemas:
- [x] Edit official docker-compose schema according to dappnode requirements. Take into account safekeys from dappmanager
- [x] Move unsafeCompose from dappmanage to SDK
- [x] Abstract code from wizard manifest and compose for:
- [ ] Use the DAppNodeSDK as a dependency in the dappmanager. Consider using json-schema-to-ts tool to infer types. Use of `yarn link` to use this PR as a dependency
- [ ] Link docs with schemas

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/compose_feature.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+---
```
Comment (dapplion) [inline .github/ISSUE_TEMPLATE/compose_feature.md:1]: Add all ISSUE_TEMPLATE files in a separate PR

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -34,7 +34,8 @@ jobs:
     name: Publish
     runs-on: ubuntu-latest
     needs: tag
-    if: needs.tag.outputs.tag != ''
+
+    if: github.event_name == 'workflow_dispatch'
```
Comment (dapplion) [inline .github/workflows/release.yml:38]: This changes the release flow of the SDK is a critical way. Please do this change in another PR, I don't agree with this strategy initially

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 3]
```diff
@@ -28,11 +29,13 @@ export const gaBumpUpstream: CommandModule<
...
   dir = defaultDir
 }: CliGlobalOptions): Promise<void> {
-  const { manifest, format } = readManifest({ dir });
-  const compose = readCompose({ dir });
+  const manifest = readReleaseFile(ReleaseFileType.manifest, { dir });
+  const compose = readReleaseFile(ReleaseFileType.compose, { dir });
 
-  const upstreamRepos = parseCsv(manifest.upstreamRepo);
-  const upstreamArgs = parseCsv(manifest.upstreamArg || "UPSTREAM_VERSION");
+  const upstreamRepos = parseCsv(manifest.releaseFile.upstreamRepo);
```
Comment (dapplion) [inline src/commands/githubActions/bumpUpstream/index.ts:35]: It's strange to have to do `manifest.releaseFile.upstreamRepo` instead of `manifest.upstreamRepo`. In reality you have placed the Manifest type in `manifest.releaseFile`, so the variable you call `manifest` is not really the manifest but something else. What positive value is this adding? At least for my subjective opinion it adds confusion and boilerplate

File: src/params.ts [hunk 4]
```diff
@@ -115,9 +102,6 @@ export const releaseFilesDefaultNames: {
   compose: "docker-compose.yml",
   avatar: "avatar.png",
   setupWizard: "setup-wizard.json",
-  setupSchema: "setup.schema.json",
-  setupTarget: "setup-target.json",
-  setupUiJson: "setup-ui.json",
```
Comment (dapplion) [inline src/params.ts:120]: We cannot merge this change until there's hard proof that no package is using the old setup wizard v1 feature

File: src/releaseFiles/manifest/manifest.schema.json [hunk 5]
```diff
@@ -42,23 +43,46 @@
       "description": "Type of this DAppNode Package. It is used to trigger some special features such as core functionality.",
       "default": "service",
       "examples": ["service", "dncore"],
-      "enum": ["service", "library", "dncore"],
-      "enumDescriptions": {
-        "service": "Can have ENVs a depend on library DAppNode Packages",
-        "library": "Should not have ENVs and cannot depend on service DAppNode Packages",
-        "dncore": "DAppNode Packages that are part of the DAppNode core and have special permissions"
-      }
+      "anyOf": [
```
Comment (dapplion) [inline src/releaseFiles/manifest/manifest.schema.json:46]: What the issue with the `enum` syntax? The docs will probably break as they only understand the `enum` syntax. Can you double check they can render the `anyOf` as pretty as `enum`?

File: src/releaseFiles/validateSchema.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+ *   errors: [
+ *     "manifest should have required property 'description'",
+ *     "manifest.image.path should be an non-empty string",
+ *   ]
+ * }
+ */
+function validateReleaseFile(
+  releaseFile: ReleaseFile
+): { valid: boolean; errors: string[] } {
+  const ajv = new Ajv({
```
Comment (dapplion) [inline src/releaseFiles/validateSchema.ts:35]: Ajv should only be initialized once in the global scope if options are not configurable

File: src/releaseFiles/validateSchema.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    coerceTypes: true,
+    strictSchema: false
+  });
+  ajvErrors(ajv);
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  let schema: any;
+  const releaseFileType = releaseFile.type;
+
+  // Precompile validator
```
Comment (dapplion) [inline src/releaseFiles/validateSchema.ts:46]: You are not pre-compiling since you run `compile` on every call to `validateReleaseFile`

File: src/releaseFiles/validateSchema.ts [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+ *   keyword: "pattern",
+ *   dataPath: "/avatar",
+ *   schemaPath: "#/properties/avatar/pattern",
+ *   params: { pattern: "^/(ipfs|bzz)/w+$" },
+ *   message: 'should match pattern "^/(ipfs|bzz)/w+$"'
+ * }
+ * @returns errorMessage:
+ * "manifest.avatar should match pattern "^/(ipfs|bzz)/w+$""
+ */
+export function processError(
```
Comment (dapplion) [inline src/releaseFiles/validateSchema.ts:122]: This should be in another file, and could be named `renderAjvError`

**Review by dapplion — CHANGES_REQUESTED** [review]

File: .github/ISSUE_TEMPLATE/compose_feature.md (hunk 1 above)
Comment (pablomendezroyo) [reply .github/ISSUE_TEMPLATE/compose_feature.md]: done

File: .github/workflows/release.yml (hunk 2 above)
Comment (pablomendezroyo) [reply .github/workflows/release.yml]: done

File: src/releaseFiles/manifest/manifest.schema.json (hunk 5 above)
Comment (pablomendezroyo) [reply src/releaseFiles/manifest/manifest.schema.json:46]: this is not a change from this PR. The manifest and wizard schemas exist already, they were just replaced. The new one is the compose schema which is the official one

The docs are not broken and understand the syntax
https://docs.dappnode.io/developers/manifest-reference#chain

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 3 above)
Comment (pablomendezroyo) [reply src/commands/githubActions/bumpUpstream/index.ts]: changed to 
```
{ data: ... , format: ...}
``` 

I think this is way more accuarate. The structure is needed to be able to abstract the function that reads, writes and validates. I think it worths in order to simplify code

**Comment by pablomendezroyo** [comment]
closed by https://github.com/dappnode/DAppNodeSDK/pull/237 https://github.com/dappnode/DAppNodeSDK/pull/236 https://github.com/dappnode/DAppNodeSDK/pull/229 and https://github.com/dappnode/DAppNodeSDK/pull/223
